### ISSUE TITLE
How to alter the type of field?

### ISSUE THREAD

**Issue by zl2fxy** [OrchardCMS/OrchardCore#2783]
hi，
If I want to amend fieldtype，how will i do it，
Using  OfType(),like as it,but it seems useless.
i want to alter but to add field，it is just useful for adding field.   
_contentDefinitionManager.AlterPartDefinition(model.PartName, partBuilder =>
               {
                      partBuilder.WithField(model.Name, fieldBuilder =>
                        {
                                 fieldBuilder.OfType(model.Type);
                                 fieldBuilder.WithDisplayName(model.DisplayName);
                       });
               });
thank you!

**Comment by zl2fxy** [comment]
I don't know why not？
because if field alter fieldtype，field data shouldn't display in the page correctly？

**Comment by Skrypt** [comment]
Yes, you shouldn't alter a field after you created content items with it. If, for example you would alter that field from Textfield to Boolean ; it's data stored in the database will be different. So, you should create a new field and then migrate data from the old field to the new one by using type conversion methods.

**Comment by zl2fxy** [comment]
> Yes, you shouldn't alter a field after you created content items with it. If, for example you would want to alter that field from Textfield to Boolean ; it's data stored in the database would be different from one type to the other. So, in that case I think you should create a new field and then migrate the data from the old field to the new one by using type conversion methods.

Ok,Thank you,but if If an ordinary user to use Cms,he has many data for a field,then he delete the field,and created a different type field that they has same Technical Name,so the Contentitem will have a empty field data on displaying.
Now I have a idea,we user same key when we to store data，we pull data  by using type conversion methods。
for example
textfield
 public override IDisplayResult Edit(TextField field, BuildFieldEditorContext context)
        {
            return Initialize<EditTextFieldViewModel>(GetEditorShapeType(context), model =>
            {
                model.Value = field.Value.Tostring();
                model.Field = field;
                model.Part = context.ContentPart;
                model.PartFieldDefinition = context.PartFieldDefinition;
            });
        }

NumericField
public override IDisplayResult Edit(NumericField field, BuildFieldEditorContext context)
        {
            return Initialize<EditNumericFieldViewModel>(GetEditorShapeType(context), model =>
            {
                var settings = context.PartFieldDefinition.Settings.ToObject<NumericFieldSettings>();
                model.Value =int.parse(context.IsNew ? settings.DefaultValue : Convert.ToString(field.Value, CultureInfo.CurrentUICulture));

                model.Field = field;
                model.Part = context.ContentPart;
                model.PartFieldDefinition = context.PartFieldDefinition;
            });
        }

**Comment by sebastienros** [comment]
Let me rephrase the issue. You want to be able to change the types of field, and if the physical types are compatible then you would expect the data to be kept intact. For instance, converting from Html to Text, or Markdown to Html, which are store as strings, would be allowed. 

Unfortunately that would be a huge breaking change for OC as we would have to use the same property names based on a physical type, so we can't do it. And it would also make things harder to use, as instead of typing `myField.Markdown` we would type `myField.Text` so we'd lose some information.

The best solution is to do a script that will process all content items and change their state.
That could be something we would provide, I assume lots of scenarios might require content items migrations like this. We would provide a javascript script that would convert the content items.

Another option is to use the Activating event which will be triggered when a content item is loaded from the content manager. You could update the inner content at that point.

I will close this issue as there is nothing directly actionable, but will create another issue to provide a migration utility.

https://github.com/OrchardCMS/OrchardCore/issues/2808
